# RNAseqEval
A collection of tools for evaluating RNA seq mapping

## generate_transcriptome.py
Run generate_transcriptome.py to generate a trancriptome from a reference and a set of annotations.

How to run:
  
     generate_transcriptome.py annotations.gff reference.fasta transcriptome.fasta


## RNAseqEval
Run RNAseqEval.py for general evaulation of mappings in sam file against reference and optionally annotations. Run RNAseqEval.py without any arguments to print options.

Usage:
     
    RNAseqEval.py eval-mapping <reference FASTA file> <input SAM file> options

    options:
    -a <file> : a reference annotation (GFF/GTF/BED) file
    -o (--output) <file> : output file to which the report will be written


## Process_pbsim_data.py
Run Process_pbsim_data.py to evaluate mappings of data generated by pbsim simulator.

How to run:
 
    Process_pbsim_data.py process simulation_root_folder mappings.sam annotations.bed

The script is written so that it can evaluate a file obtained from several simulations. Pbsim output folders must be within the simulation_root_folder, and a dictionary simFolderDict should be modified within the script to connect each simualted read header with the corresponding pbsim output folder.
